Saving an ItemStack results in an NullPointerException

```java
import java.io.File;

import org.bukkit.Material;
import org.bukkit.inventory.ItemStack;

import net.cubespace.Yamler.Config.Config;

public class TestConfig extends Config
{
	public ItemStack is = new ItemStack(Material.COOKIE);

    public TestConfig()
    {
        CONFIG_FILE = new File("test.yml");
        try
        {
            addConverter(net.cubespace.Yamler.Converter.ItemStack.class);
            init();
        } catch (Exception e)
        {
            e.printStackTrace();
        }
    }

    public static void main(String args[])
    {
        new TestConfig();
    }
}
```

This produces the following NullPointerException:
```java
net.cubespace.Yamler.Config.InvalidConfigurationException: Could not save the Field
	at net.cubespace.Yamler.Config.Config.internalSave(Config.java:84)
	at net.cubespace.Yamler.Config.Config.save(Config.java:36)
	at net.cubespace.Yamler.Config.Config.init(Config.java:108)
	at TestConfig.<init>(TestConfig.java:18)
	at TestConfig.main(TestConfig.java:27)
Caused by: java.lang.NullPointerException
	at org.bukkit.Bukkit.getItemFactory(Bukkit.java:1038)
	at org.bukkit.inventory.ItemStack.getItemMeta(ItemStack.java:568)
	at net.cubespace.Yamler.Converter.ItemStack.toConfig(ItemStack.java:34)
	at net.cubespace.Yamler.Config.InternalConverter.toConfig(InternalConverter.java:102)
	at net.cubespace.Yamler.Config.Config.internalSave(Config.java:80)
	... 4 more
```

--- a/Yamler-Bukkit/src/main/java/net/cubespace/Yamler/Converter/ItemStack.java
+++ b/Yamler-Bukkit/src/main/java/net/cubespace/Yamler/Converter/ItemStack.java
@@ -30,9 +30,12 @@ public class ItemStack implements Converter {
 
         Converter listConverter = converter.getConverter(List.class);
 
-        Map<String, Object> meta = new HashMap<>();
-        meta.put("name", itemStack.getItemMeta().hasDisplayName() ? itemStack.getItemMeta().getDisplayName() : null);
-        meta.put("lore", itemStack.getItemMeta().hasLore() ? listConverter.toConfig(List.class, itemStack.getItemMeta().getLore(), null) : null);
+        Map<String, Object> meta = null;
+        if(itemStack.hasItemMeta()) {
+            meta = new HashMap<>();
+            meta.put("name", itemStack.getItemMeta().hasDisplayName() ? itemStack.getItemMeta().getDisplayName() : null);
+            meta.put("lore", itemStack.getItemMeta().hasLore() ? listConverter.toConfig(List.class, itemStack.getItemMeta().getLore(), null) : null);
+        }
 
         saveMap.put("meta", meta);
 
@@ -41,24 +44,36 @@ public class ItemStack implements Converter {
 
     @Override
     public Object fromConfig(Class type, Object section, ParameterizedType genericType) throws Exception {
-        Map<String, Object> itemstackMap = (Map<String, Object>) ((ConfigSection) section).getRawMap();
-        Map<String, Object> metaMap = (Map<String, Object>) ((ConfigSection) itemstackMap.get("meta")).getRawMap();
+        Map itemstackMap;
+        Map metaMap = null;
+
+        if(section instanceof Map) {
+            itemstackMap = (Map) section;
+            metaMap = (Map) itemstackMap.get("meta");
+        } else {
+            itemstackMap = ((ConfigSection) section).getRawMap();
+            if(itemstackMap.get("meta") != null)
+                metaMap = ((ConfigSection) itemstackMap.get("meta")).getRawMap();
+        }
 
         String[] temp = ((String) itemstackMap.get("id")).split(":");
-        org.bukkit.inventory.ItemStack itemStack = new org.bukkit.inventory.ItemStack(Material.valueOf(temp[0]));
+        org.bukkit.inventory.ItemStack itemStack = (org.bukkit.inventory.ItemStack)
+                type.getConstructor(Material.class).newInstance(Material.valueOf(temp[0]));
         itemStack.setAmount((int) itemstackMap.get("amount"));
 
-        if (temp.length == 2) {
+        if(temp.length == 2) {
             itemStack.setDurability(Short.valueOf(temp[1]));
         }
 
-        if (metaMap.get("name") != null) {
-            itemStack.getItemMeta().setDisplayName((String) metaMap.get("name"));
-        }
+        if(metaMap != null) {
+            if(metaMap.get("name") != null) {
+                itemStack.getItemMeta().setDisplayName((String) metaMap.get("name"));
+            }
 
-        if (metaMap.get("lore") != null) {
-            Converter listConverter = converter.getConverter(List.class);
-            itemStack.getItemMeta().setLore((List<String>) listConverter.fromConfig(List.class, metaMap.get("lore"), null));
+            if(metaMap.get("lore") != null) {
+                Converter listConverter = converter.getConverter(List.class);
+                itemStack.getItemMeta().setLore((List<String>) listConverter.fromConfig(List.class, metaMap.get("lore"), null));
+            }
         }
 
         return itemStack;

--- a/Yamler-Bukkit/src/test/java/base/BaseTest.java
+++ b/Yamler-Bukkit/src/test/java/base/BaseTest.java
@@ -1,0 +1,27 @@
+package base;
+
+import net.cubespace.Yamler.Config.Config;
+
+import java.io.File;
+
+/**
+ * @author geNAZt (fabian.fassbender42@googlemail.com)
+ */
+public abstract class BaseTest {
+    protected String filename;
+    protected Config config;
+    protected File file;
+
+    public void before() throws Exception {
+        file = new File("temp", filename);
+        if(!file.getParentFile().exists()) {
+            file.getParentFile().mkdirs();
+        }
+
+        if(file.exists()) {
+            file.delete();
+        }
+    }
+
+    public abstract void setup() throws Exception;
+}

--- a/Yamler-Bukkit/src/test/java/converter/ItemStackConverterTest.java
+++ b/Yamler-Bukkit/src/test/java/converter/ItemStackConverterTest.java
@@ -1,0 +1,44 @@
+package converter;
+
+import base.BaseTest;
+import converter.config.ItemStackTestConfig;
+import net.cubespace.Yamler.Config.InvalidConfigurationException;
+import net.cubespace.Yamler.Converter.ItemStack;
+import org.testng.annotations.BeforeSuite;
+import org.testng.annotations.Test;
+
+/**
+ * A basic ItemStack converter test to ensure the proper saving and
+ * loading of a ItemStack.
+ *
+ * @author bibo38
+ * @see ItemStackTestConfig
+ * @see ItemStack
+ */
+public class ItemStackConverterTest extends BaseTest {
+    /**
+     * The setup method, which creates the configuration file
+     * and the Configuration used for the following tests.
+     *
+     * @throws Exception
+     */
+    @Override
+    @BeforeSuite
+    public void setup() throws Exception {
+        config = new ItemStackTestConfig();
+        filename = "itemstackConverterTest.yml";
+
+        before();
+    }
+
+    /**
+     * A basic test to ensure the proper loading and saving of the Config.
+     *
+     * @throws InvalidConfigurationException If an configuration error occurs
+     */
+    @Test
+    public void testInitLoad() throws InvalidConfigurationException {
+        config.init(file);
+        config.load();
+    }
+}

--- a/Yamler-Bukkit/src/test/java/converter/config/ItemStackTestConfig.java
+++ b/Yamler-Bukkit/src/test/java/converter/config/ItemStackTestConfig.java
@@ -1,0 +1,111 @@
+package converter.config;
+
+import helper.TestItemStack;
+import net.cubespace.Yamler.Config.Config;
+import org.bukkit.Material;
+import org.bukkit.enchantments.Enchantment;
+import org.bukkit.inventory.meta.ItemMeta;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * The test configuration to check for a correct working ItemStack.
+ *
+ * @author bibo38
+ * @see converter.ItemStackConverterTest
+ */
+public class ItemStackTestConfig extends Config {
+
+    public TestItemStack cookie = new TestItemStack(Material.COOKIE);
+    public TestItemStack specialWool = new TestItemStack(Material.WOOL);
+
+    /**
+     * The default constructor is used to add the {@link net.cubespace.Yamler.Converter.ItemStack} converter
+     * and add a special ItemMeta class to the {@link this#specialWool} variable to test the converter.
+     *
+     * @throws Exception When the converter cannot be added
+     */
+    public ItemStackTestConfig() throws Exception {
+        addConverter(net.cubespace.Yamler.Converter.ItemStack.class);
+
+        specialWool.setItemMeta(new ItemMeta() {
+            @Override
+            public boolean hasDisplayName() {
+                return true;
+            }
+
+            @Override
+            public String getDisplayName() {
+                return "Test Wool";
+            }
+
+            @Override
+            public void setDisplayName(String s) {
+            }
+
+            @Override
+            public boolean hasLore() {
+                return true;
+            }
+
+            @Override
+            public List<String> getLore() {
+                return Arrays.asList("This is Wool", "Good Wool");
+            }
+
+            @Override
+            public void setLore(List<String> list) {
+            }
+
+            @Override
+            public boolean hasEnchants() {
+                return true;
+            }
+
+            @Override
+            public boolean hasEnchant(Enchantment enchantment) {
+                return Enchantment.LUCK.equals(enchantment);
+            }
+
+            @Override
+            public int getEnchantLevel(Enchantment enchantment) {
+                return hasEnchant(enchantment) ? 1 : 0;
+            }
+
+            @Override
+            public Map<Enchantment, Integer> getEnchants() {
+                HashMap<Enchantment, Integer> enchantments = new HashMap<>();
+                enchantments.put(Enchantment.LUCK, 1);
+                return enchantments;
+            }
+
+            @Override
+            public boolean addEnchant(Enchantment enchantment, int i, boolean b) {
+                return false;
+            }
+
+            @Override
+            public boolean removeEnchant(Enchantment enchantment) {
+                return false;
+            }
+
+            @Override
+            public boolean hasConflictingEnchant(Enchantment enchantment) {
+                return false;
+            }
+
+            @Override
+            public ItemMeta clone() {
+                return null;
+            }
+
+            @Override
+            public Map<String, Object> serialize() {
+                return null;
+            }
+        });
+    }
+}

--- a/Yamler-Bukkit/src/test/java/helper/TestItemStack.java
+++ b/Yamler-Bukkit/src/test/java/helper/TestItemStack.java
@@ -1,0 +1,121 @@
+package helper;
+
+import org.bukkit.Material;
+import org.bukkit.enchantments.Enchantment;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.inventory.meta.ItemMeta;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * An implementation of the default Bukkit {@link ItemStack} to provide
+ * a similar ability to write/read the {@link ItemMeta} without CraftBukkit running.
+ *
+ * @author bibo38
+ * @see ItemStack
+ */
+public class TestItemStack extends ItemStack {
+    ItemMeta meta = null;
+
+    /**
+     * The common used constructor
+     *
+     * @param type The type of the ItemStack
+     * @see ItemStack#ItemStack(Material)
+     */
+    public TestItemStack(Material type) {
+        super(type);
+    }
+
+    @Override
+    public ItemMeta getItemMeta() {
+        if(meta == null)
+            meta = new ItemMeta() {
+                @Override
+                public boolean hasDisplayName() {
+                    return false;
+                }
+
+                @Override
+                public String getDisplayName() {
+                    return null;
+                }
+
+                @Override
+                public void setDisplayName(String s) {
+                }
+
+                @Override
+                public boolean hasLore() {
+                    return false;
+                }
+
+                @Override
+                public List<String> getLore() {
+                    return null;
+                }
+
+                @Override
+                public void setLore(List<String> list) {
+                }
+
+                @Override
+                public boolean hasEnchants() {
+                    return false;
+                }
+
+                @Override
+                public boolean hasEnchant(Enchantment enchantment) {
+                    return false;
+                }
+
+                @Override
+                public int getEnchantLevel(Enchantment enchantment) {
+                    return 0;
+                }
+
+                @Override
+                public Map<Enchantment, Integer> getEnchants() {
+                    return null;
+                }
+
+                @Override
+                public boolean addEnchant(Enchantment enchantment, int i, boolean b) {
+                    return false;
+                }
+
+                @Override
+                public boolean removeEnchant(Enchantment enchantment) {
+                    return false;
+                }
+
+                @Override
+                public boolean hasConflictingEnchant(Enchantment enchantment) {
+                    return false;
+                }
+
+                @Override
+                public ItemMeta clone() {
+                    return null;
+                }
+
+                @Override
+                public Map<String, Object> serialize() {
+                    return null;
+                }
+            };
+        return meta;
+    }
+
+    @Override
+    public boolean hasItemMeta() {
+        return meta != null;
+    }
+
+    @Override
+    public boolean setItemMeta(ItemMeta itemMeta) {
+        meta = itemMeta;
+        return true;
+    }
+}

